keps/sig-node/20190412-lxcfs.md

```diff
@@ -0,0 +1,189 @@
+- enable pod has a nice resource view.
+
+### Non-Goals
+
+This feature only mount lxcfs files, the implemention of resource view is depend on [lxcfs](https://github.com/lxc/lxcfs)
+
+## Proposal
+
+### Implementation Details/Notes/Constraints [optional]
+
+Add lxcfs volumes when mount pod volumes:
+```golang
+const lxcfsProcPath = "/var/lib/lxcfs/proc/"
+
+var lxcfsProcEntry = []string{"cpuinfo", "meminfo", "diskstats", "uptime", "stat", "swaps"}
+
+//LxcfsMounts mount lxcfs files
+func LxcfsMounts() []container.Mount {
+	var mounts []container.Mount
+	for _, entry := range lxcfsProcEntry {
+		mount := container.Mount{
+			Name:           fmt.Sprintf("lxcfs-%s", entry),
+			ContainerPath:  fmt.Sprintf("/proc/%s", entry),
+			HostPath:       lxcfsProcPath + entry,
+			ReadOnly:       true,
+			SELinuxRelabel: false,
+			Propagation:    runtimeapi.MountPropagation_PROPAGATION_HOST_TO_CONTAINER,
+		}
+		mounts = append(mounts, mount)
+	}
+	return mounts
+}
+```
+
+Then add a feature gate, default is closed.
+```golang
+	if utilfeature.DefaultFeatureGate.Enabled(features.KubeletSupportLxcfs) {
+		lxcfsMounts := LxcfsMounts()
+		opts.Mounts = append(opts.Mounts, lxcfsMounts...)
+	}
+```
+[More details](https://github.com/kubernetes/kubernetes/pull/75475)
+
+### Use guide
+
+> install lxcfs 
+
+```
+wget https://copr-be.cloud.fedoraproject.org/results/ganto/lxd/epel-7-x86_64/00486278-lxcfs/lxcfs-2.0.5-3.el7.centos.x86_64.rpm
+yum install lxcfs-2.0.5-3.el7.centos.x86_64.rpm 
+systemctl enable lxcfs
+systemctl start lxcfs
+```
+
+> Test 
+
+```
+mkdir /root/static
+touch  /root/static/pod.yaml
+```
+
+```
+[root@izj6c996fvahbz1keqg2iqz ~]# cat static/pod.yaml 
+apiVersion: v1
+kind: Pod
+metadata:
+  name: myapp-pod
+  labels:
+    app: myapp
+spec:
+  containers:
+  - name: myapp-container
+    image: ubuntu:18.04
+    command: ['sh', '-c', 'echo Hello Kubernetes! && sleep 3600']
+    resources:
+      requests:
+        memory: "1Gi"
+        cpu: "500m"
+      limits:
+        memory: "1Gi"
+        cpu: "500m"
+```
+
+```
+./kubelet --pod-manifest-path /root/static  --feature-gates KubeletSupportLxcfs=true
+```
+
+```
+docker exec -it xxxx bash
+
+root@myapp-pod-izj6c996fvahbz1keqg2iqz:/# free -h
+              total        used        free      shared  buff/cache   available
+Mem:           1.0G        928K        1.0G        720K          0B        1.0G
+Swap:            0B          0B          0B
+```
+As you can see, the memory is 1.0G，the host memory is 4G
+
+### Risks and Mitigations
+
+When lxcfs restart, needs remount lxcfs files
+
+## Test plan
+> Unit test
+
+Provide the unit test code in v1.5
+
+> E2e test
+
+Provide e2e test cases in v1.5
+
+## Graduation Criteria
+
+### Kubelet support lxcfs
+
+The following criteria applies to `SupportKubeletLxcfs` feature gate:
+
+Alpha
+- basic support integrated in kubelet
+
+Beta
+- ensure proper node e2e test coverage is integrated verifying cgroup settings
```

Comment:
Does this mean on by default and used everywhere?  Typically feature gates once enabled are on everywhere, and it doesn’t seem like this is universally portable.  It’s partly why I am wondering if this belongs more in the CRI provider.

Reply:
+1, as I mentioned in one of previous comments, we can not assume lxcfs is installed in every cluster. Defer to a CRI provider would make more sense. AFAIK, there're runtimes planning to implement lxcfs support while have to workaround CRI.

Reply:
Yes, some runtime like pouch already supports this feature, but our platform using multiple runtimes we feel tired to implement it on all runtimes. Now I have an other idea to solve this problem, call it kubelet premout, we can appoint some volumes in the config file and mount then when pod creating, so we needn't care about lxcfs, and it also can use in other use case.s.